common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -46,94 +46,112 @@ template <int subwarp_size, int subwarps_per_block, typename ValueType,
 __forceinline__ __device__ void generic_generate(
     IndexType num_rows, const IndexType *__restrict__ m_row_ptrs,
     const IndexType *__restrict__ m_col_idxs,
-    const ValueType *__restrict__ m_values, IndexType *__restrict__ i_row_ptrs,
-    IndexType *__restrict__ i_col_idxs, ValueType *__restrict__ i_values,
-    Callable trs_solve)
+    const ValueType *__restrict__ m_values,
+    const IndexType *__restrict__ i_row_ptrs,
+    const IndexType *__restrict__ i_col_idxs, ValueType *__restrict__ i_values,
+    IndexType *__restrict__ excess_rhs_sizes,
+    IndexType *__restrict__ excess_nnz, Callable trs_solve)
 {
-    static_assert((subwarp_size & (subwarp_size - 1)) == 0,
-                  "subwarp_size must be a power of 2");
-
-    const auto gidx = thread::get_thread_id_flat<IndexType>();
+    static_assert(subwarp_size >= row_size_limit, "incompatible subwarp_size");
     const auto row = thread::get_subwarp_id_flat<subwarp_size, IndexType>();
-    const int tidx = threadIdx.x;
+
+    if (row >= num_rows) {
+        return;
+    }
+
+    const auto i_row_begin = i_row_ptrs[row];
+    const auto i_row_size = i_row_ptrs[row + 1] - i_row_begin;
 
     auto this_block = group::this_thread_block();
-    auto subwarp = group::tiled_partition<subwarp_size>(this_block);
+    auto subwarp =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
     const int local_id = subwarp.thread_rank();
 
-    __shared__ UninitializedArray<ValueType, subwarp_size * subwarp_size *
-                                                 subwarps_per_block>
-        trisystem;
+    if (i_row_size > subwarp_size) {
+        // defer long rows: store their nnz and number of matches
+        IndexType count{};
+        for (IndexType nz = 0; nz < i_row_size; ++nz) {
+            auto col = i_col_idxs[i_row_begin + nz];
+            auto m_row_begin = m_row_ptrs[col];
+            auto m_row_size = m_row_ptrs[col + 1] - m_row_begin;
+            // extract the sparse submatrix consisting of the entries whose
+            // columns/rows match column indices from this row
+            group_match<subwarp_size>(
+                m_col_idxs + m_row_begin, m_row_size, i_col_idxs + i_row_begin,
+                i_row_size, subwarp,
+                [&](IndexType, IndexType, IndexType,
+                    config::lane_mask_type matchmask,
+                    bool) { count += popcnt(matchmask); });
```

Comment:
this only count the number of nonzero, right?

Reply:
Yes, this counts the number of non-zeros in the sparse triangular submatrix with row/column indices matching the column indices from the current row.

Reply:
thanks